In [73]:
import pandas as pd
import numpy as np


## Reading files

In [62]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [5]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [6]:
wh = pd.concat([wh2018,wh2019])

In [7]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [8]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [10]:
for c_ in wh.columns:
    print(c_)
    print(np.sum(wh[c_].isna()))

Overall rank
0
Country or region
0
Score
19
GDP per capita
12
Social support
16
Healthy life expectancy
26
Freedom to make life choices
23
Generosity
21
Perceptions of corruption
13
year
0


In [32]:
porcentaje =((wh.isnull().sum() / len(wh))*100).sort_values(ascending=False)

In [38]:
porcentaje

porcentaje                      100.000000
Healthy life expectancy           8.333333
Freedom to make life choices      7.371795
Generosity                        6.730769
Score                             6.089744
Social support                    5.128205
Perceptions of corruption         4.166667
GDP per capita                    3.846154
Overall rank                      0.000000
Country or region                 0.000000
year                              0.000000
dtype: float64

In [57]:
a={
     "year": ["year"]
    , "column" : ["year"]
    , "percentage.of.nans" : "porcentaje"
  
}

In [58]:
pd.DataFrame(a)

year column percentage.of.nans
0  year   year         porcentaje

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [13]:
wh['year'].describe()

count     312.000000
mean     2018.500000
std         0.500803
min      2018.000000
25%      2018.000000
50%      2018.500000
75%      2019.000000
max      2019.000000
Name: year, dtype: float64

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

In [ ]:
wh.fillna(0).mean

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [62]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro

In [ ]:
wh[0;2]-wh[1;2]

In [ ]:
wh.drop(['year'],axis=1)